### DISCUSSION TITLE
Add xapp status icon for drive dirty/writeback cache

### DISCUSSION THREAD

**Discussion by billyswong** [linuxmint/discussions#429] [Ideas]
From time to time, there are users who copy too many files to a removable drive. Then when they want to unplug it, they find themselves stuck in a long wait with no idea when the sync will finish. This kind of issues has appeared in Linux Mint Forums a number of times. Some users may get impatient and unplug prematurely. Some poor users may not even know they are not ready to unplug as the indication of "things not ready" is often not clear enough.

For this problem, I came across a script suggestion at https://www.kaper.com/notes/linux-sync-buffer-progress-monitoring-e-g-for-copy-to-usb/
```
watch "grep -e Dirty: -e Writeback: /proc/meminfo"
```
Yes, such script will help a knowledgeable user to monitor the amount of data still in the cache and plan accordingly. But it still takes a steep learning curve for users who know nothing about disk cache mechanism. However primitive that is, if we can wrap such information into a more user-friendly format and present it automatically as an xapp status icon, this will sure help many people running Linux Mint.

Another feature we may integrate to this status icon is to let users click it to fire the `sync` command. Users will no longer need to type command in terminal just for a routine removable drive handling.

**Comment by spamegg1** [comment]
I second the proposal.
I've also ran into... way too many forum posts about this.
People keep losing their files all the time.
It would be nice if `sync` was pre-built into and automated for file transfers that are above a threshold of total amount / total number of files, maybe?

**Comment by Hawaiihemd** [comment]
I support this idea!

This exact problem is one of my very few gripes I do have in Mint, having set up a handful of PCs for my relatives.

While it is easy to blame the user if they do not use the eject button at all, it is IMHO really hard to expect the user to wait - sometimes minutes - for any kind of feedback after clicking to eject.

I think some kind of notification should be presented to the user. If it would not only say "trying to unmount safely, please wait..." but would include even a rough estimate based on the code billyswong mentioned above, that would be all one could ask for.

**Comment by ost-guy-comeback** [comment]
# I support 

because sometimes a guy unplug the drive but click on the safely unplug and a iso is broken because of that and seriously we need better usb management

**Comment by cplittlfield** [comment]
Cinnamon users can install the Sync applet (sync@szlldm).  I use it, and it's very helpful.  Don't MATE and Xfce have similar features?

**Comment by mtwebster** [comment]
There should be a notification when a device shouldn't be removed yet. There has been quite a bit of work on this as well for the next Cinnamon release: 
https://github.com/linuxmint/cinnamon/commit/9bdd28c5930b894a8d8e2ea0eac62e80d2754df5

A status applet for file cache state is not the answer here. Making sure places that allow the user to eject drives function properly is.

**Reply by billyswong** [reply]
There is likely some miscommunication here...

1. A Cinnamon-only solution doesn't cover MATE and Xfce. (But I won't complain if you guys want to create a Cinnamon-only solution and leave the matter of MATE and Xfce to their own teams. I just think this feature can be DE-agnostic.)
2. Cache state matters because too often routine file operations claim the file copy/move operations have finished in GUI but there is a huge amount of data still waiting to be written from the cache. The unmount / eject command GUI often only tells people things are not ready yet, but lack a progress indication and no obvious way to see how long one should expect to wait. Without such progress indication, no static wording can get impatient users to stop their reckless unplug. I've seen this behaviour in Windows users too often.
3. Users rightfully expect the unmount / eject command to take a short and fixed amount of time to complete. The hidden "sync" operation deserves its own progress GUI. Anything that takes a variable length of time + can be over 1 minute long (some poor cases may take hours) deserve their own progress GUI.
